### PR TITLE
Add LangSmith source

### PR DESCRIPTION
**Pull request by dsmilkov**
- Allow sources to register their own server-side API handlers (e.g. pinging the LangSmith server for a list of valid datasets)
- Fix a bug with HNSW index where datasets smaller than default K (20) were not supported

https://github.com/lilacai/lilac/assets/2294279/ed5bfb82-b914-4839-9a74-6447392fc7e2

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -70,6 +70,9 @@ regex = "^2023.6.3"
 # For language detection.
 langdetect = { version = "^1.0.9", optional = true }
 
+# Langsmith source ingestion.
+langsmith = { version = "^0.0.29", optional = true }
```
Comment (nsthorat) [inline pyproject.toml:74]: should we add another extra for all sources?

File: web/blueprint/src/lib/queries/langsmithQueries.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import {LangsmithService} from '$lilac';
+import {createApiQuery} from './queryUtils';
+
+const TAG = 'langsmith';
```
Comment (nsthorat) [inline web/blueprint/src/lib/queries/langsmithQueries.ts:4]: LANGSMITH_TAG for consistency

**Review by nsthorat — APPROVED** [review]

File: lilac/sources/langsmith.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  _source_schema: Optional[SourceSchema] = None
+  _items: Optional[Item] = None
+
+  @override
+  def setup(self) -> None:
+    try:
+      from langsmith import Client
+    except ImportError:
+      raise ImportError('Could not import dependencies for the LangSmith source. '
+                        'Please install the optional dependency via `pip install langsmith`.')
```
[the remark this answers is not in the document]
Comment (dsmilkov) [reply lilac/sources/langsmith.py]: done

File: pyproject.toml (hunk 1 above)
Comment (dsmilkov) [reply pyproject.toml:74]: done

File: web/blueprint/src/lib/queries/langsmithQueries.ts (hunk 2 above)
Comment (dsmilkov) [reply web/blueprint/src/lib/queries/langsmithQueries.ts:4]: done

**Pull request merged**